ROADMAP.md

```diff
@@ -2,23 +2,100 @@
-  - We want pull out as much non-koop specific code as possible into separate modules
-* Add support for the simple style spec - https://github.com/mapbox/simplestyle-spec
-* Address consistency issues in method calls to the Cache from providers 
-  - be clear and consitent about passing table names instead of "types" and "keys"
-
-## September 2015
-
-### Changes 
-* Explore removing ogr2ogr as a dep for exporting file formats
-### Additions 
-* Support for exporting FGDB
-* Support for OGC WMS and WFS 
-* A koop desktop app built with something like Electron
+## v3
+
+Major API changes:
+
+Provide better support for `plugins`. Right now the model for creating a plugin is pretty poor, as illustrated by https://github.com/koopjs/koop/issues/195.
+
+Move away from `provider` in favor of three new types of plugins: `source`, `export`, and `interface`.
+
+This can be thought of more or less like so:
+
+`server/koop/:source{/**}/:item{.:export,/:interface}`
+
+### Source
+
+Defines methods for extracting data from a third party source and transforming it into standard GeoJSON. Handles registering hosts if necessary.
+
+This definition is looser than a provider in that it encompasses any kind of data source, not just open data providers.
+
+Ideally a `source` plugin is more focused, meaning that
+
+- (b) it is only concerned with the specifics of its data source and how to hand that data over to koop
+- (c) it does not interact directly with the koop cache in any way
+- (a) it is less tightly coupled to express and has as little knowledge of server requests and responses as necessary
+
+Addresses:
+
+* `lib/` refactor
+* better encapsulation
+* BaseModel & BaseController issues & limitations
+
+Module naming convention:
+
+```
+koop-source-*
+
+e.g.
+
+koop-source-agol
+koop-source-socrata
+koop-source-ckan
+```
+
+Methods:
+
+* extract (internal method, can be used by itself, e.g. `require(koop-source-agol/extract)`)
+* transform (internal method, can be used by itself, e.g. `require(koop-source-agol/transform)`)
+* fetch (primary method used by koop core to request data from source)
+
+### Export
+
+Defines methods for exporting GeoJSON to a static format. This means KML, CSV, Shapefile, etc. This plugin would exist exclusively to encapsulate logic for downloading data in a specific format.
+
+Addresses:
+
+* Support for TopoJSON ([#199](https://github.com/koopjs/koop/issues/199))
+* File Geodatabase support
+* Removing ogr2ogr as a dependency for exporting file formats
+* `lib/` refactor
+
+Module naming convention:
+
+```
+koop-export-*
+
+e.g.
+
+koop-export-geojson
+koop-export-topojson
+koop-export-shapefile
+koop-export-kml
+koop-export-csv
+```
+
+### Interface
+
+Defines methods for querying GeoJSON via a third-party API. This means Feature Service, Soda2, OData, etc. This requires a lot of refactoring but would yield a lot of benefits and has the potential to set a standard interface for querying GeoJSON.
```

Comment:
I agree it's a very big change. The idea is mainly in response to trying to find a way to serve the use case @timwis brought up in [#195](https://github.com/koopjs/koop/issues/195) (How to build other API output formats?). This does raise the question of whether we should aim to support other API output formats. Is this something we want Koop to do?

I think starting with modularizing Koop's feature service querying and output is an excellent idea.